File: src/server/model/Instrument.cpp
#include "Instrument.h"
#include "base/common/Time.h"

void Instrument::Serialize(Proto::Instrument *inst) const {
  inst->set_id(id_);
  inst->set_symbol(symbol_);
  inst->set_product(product_);
  inst->set_exchange(exchange_);
  inst->set_type(type_);
  inst->set_currency(currency_);
  inst->set_lot(lot_);
  inst->set_tick(tick_);
  inst->set_multiplier(multiplier_);
  if (underlying_) {
    inst->set_underlying(underlying_->Id());
  }
  if (hedge_underlying_) {
    inst->set_hedge_underlying(hedge_underlying_->Id());
  }
  inst->set_highest(highest_);
  inst->set_lowest(lowest_);
  inst->set_commission(commission_type_);
  inst->set_open_commission(open_commission_);
  inst->set_close_commission(close_commission_);
  inst->set_close_today_commission(close_today_commission_);
  inst->set_maturity(base::DateToString(maturity_));
  inst->set_status(status_);
}

File: src/client/Models/Instrument.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace client.Models
{
    public class Instrument
    {
        //public Instrument(Proto.Instrument inst)
        //{
        //    Id = inst.Id;
        //    Symbol = inst.Symbol;
        //    Exchange = inst.Exchange;
        //    Type = inst.Type;
        //    Currency = inst.Currency;
        //    Tick = inst.Tick;
        //    Multiplier = inst.Multiplier;
        //    Highest = inst.Highest;
        //    Lowest = inst.Lowest;
        //    try
        //    {
        //        Maturity = DateTime.ParseExact(inst.Maturity, "yyyyMMdd", CultureInfo.InvariantCulture);
        //    }
        //    catch { }
        //}

        public string Id { get; set; }
        public string Symbol { get; set; }
        public Proto.Exchange Exchange { get; set; }
        public Proto.InstrumentType Type { get; set; }

        private object statusMutex = new object();
        private Proto.InstrumentStatus status;
        public Proto.InstrumentStatus Status
        {
            get
            {
                lock (this.statusMutex)
                {
                    return status;
                }
            }
            set
            {
                lock (this.statusMutex)
                {
                    status = value;
                }
            }
        }

        public Proto.Currency Currency { get; set; }
        public double Tick { get; set; }
        public double Multiplier { get; set; }

        private object priceMutex = new object();
        private double highest;
        public double Highest
        {
            get
            {
                lock (this.priceMutex)
                {
                    return highest;
                }
            }
            set
            {
                lock (this.priceMutex)
                {
                    highest = value;
                }
            }
        }

        private double lowest;
        public double Lowest
        {
            get
            {
                lock (this.priceMutex)
                {
                    return lowest;
                }
            }
            set
            {
                lock (this.priceMutex)
                {
                    lowest = value;
                }
            }
        }

        public DateTime Maturity { get; set; }

        public Instrument Underlying { get; set; }
        public Instrument HedgeUnderlying { get; set; }
        public string Product { get; set; }
    }
}
